### ISSUE TITLE
Problem with contao/core package

### ISSUE THREAD

**Issue by qzminski** [contao-community-alliance/composer-plugin#44]
Contao version 3.5.2

I have a huge problem with the contao/core package being constanly installed on Windows Server. Digging into the details, when I initially run the composer update command the contao/core dependencies are being downloaded and afterwards I get an exception:

https://github.com/contao-community-alliance/composer-plugin/blob/master/src/ContaoCommunityAlliance/Composer/Plugin/Plugin.php#L400-L401

So far so good it would seem. However when I run the composer update again, the contao/core package is downloaded and installed. Afterwards when I open the package manager in the backend, the contao/core package has the ```project``` label whereas it should be a ```metapackage```.

Interestingly I've just ran the composer update on the unix server and the contao/core has been installed as well - this definitely was not the case previously. After that I have updated Contao to 3.5.3, ran the update again but contao/core was not removed.

Can you reproduce it? Is this a composer plugin bug? Or maybe a bug in the composer itself? Is it related to the installed Contao version - if so, why there is no warning or exception?

@aschempp

**Comment by discordier** [comment]
I have no clue why this is getting installed on your site. It was definately working so far. However you are the second person within one week where this happens. Therefore there might have been a change in composer itself. I was unable to reproduce this issue so far... any more details you could provide would be nice.

**Comment by aschempp** [comment]
We have tried to debug this for several hours now. The plugin is correctly injecting the Contao version and requirements, but somehow the file is still being downloaded. We also found a section in the manual saying that a metapackage does not need a dist URL, but that did not seem to make a difference.

**Comment by discordier** [comment]
Could you try with an old(er) composer.phar?

**Comment by qzminski** [comment]
I have rolled back the composer.phar update on unix server and contao/core was not removed but I am not sure if it would be installed again though.

**Comment by qzminski** [comment]
Right now I have installed Contao 3.5.0 and apparently the issue is there - contao/core with its dependencies is installed. The contao/core package is marked as **project**. Re-run does not change anything and there is no warning/exception thrown.

However, on Contao 3.5.3 the contao/core is **not** installed but its dependencies are. I do get the warning:

```Warning: Contao core was about to get installed but has been found in project root, to recover from this problem please restart the operation```

After the re-run contao/core is **not** installed and its dependencies are removed. The contao/core package is marked as **metapackage**.

Looks like if the Contao version is older than the newest one it installs the most recent one in the ```composer/vendor``` folder.

**Comment by aschempp** [comment]
did it install Contao 3.5.3 in the vendor folder when you had 3.5.0 in the root?

**Comment by qzminski** [comment]
Yes it did install Contao 3.5.3 in the ```composer/vendor``` folder. I checked the ```constnants.php``` file and it is clearly 3.5.3

**Comment by discordier** [comment]
Ok, I have now 3 People experiencing this issue of which 2 have upgraded their Contao core (via liveupdate) and you are opposing as you have an outdated version.

Therefore I wonder if it does have anything to do with having upgraded/downgraded or rather the issue is simply due to having a core version different than the one "it had before" and therefore composer tries to install that one back again.

**Comment by qzminski** [comment]
Hmm that's weird because I have now performed the composer installation on serveral 3.5.3 systems and it did **not** download contao/core. However I did not update those systems via LiveUpdate.

During the debugging with @aschempp we discovered that during the first update the local package of contao/core was type of ```metapackage``` and the exception was thrown. However upon second update, the contao/core was downloaded and the package has changed its type to ```project```. We have also checked the versions comparison and it has (afair) always been ```3.5.2.0 == 3.5.2.0```.

https://github.com/contao-community-alliance/composer-plugin/blob/master/src/ContaoCommunityAlliance/Composer/Plugin/Plugin.php#L233-L237

We have also created a fork of the plugin with some debugging but still contao/core was always installed although the events were (at least in theory) triggered before packages being installed.

https://github.com/qzminski/composer-plugin

To sum up my observations are as follow: if the installed Contao version is the newest one then contao/core is not being downloaded, otherwise it is. Unfortunately I have no idea how to do useful debugging to provide more info.

If this issue did not occur previously then maybe composer itself has introduced some changes that break the current functionality? There were some new commits in last days but looking at them I can't think of any that could break it.

https://github.com/composer/composer/commits/master

**Comment by discordier** [comment]
@qzminski Could you provide the composer.lock and installed.json or at least the version numbers from within there?

**Comment by Ainschy** [comment]
Hier die composer.lock und die installed.json

**Comment by qzminski** [comment]
Raw Contao 3.5.0 system with composer plugin installed:

https://gist.github.com/qzminski/33bdf61af24fb979daac

**Comment by discordier** [comment]
Mhhh... very strange... I am still unable to reproduce it...

**Comment by qzminski** [comment]
Shall I create a testing environment for you?

**Comment by Ainschy** [comment]
i have rename: composer/vendor in composer/vendor_old then "update packages" ... 

Then came a Warning the he has found contao in the root dir. After i push the "update packages" again all looks fine.

Sry for my english

**Comment by discordier** [comment]
@frontendschlampe and I are on it right now.

the problem does not exists with composer/composer@80f1e43

So the problem is in between https://github.com/composer/composer/compare/80f1e43...7a9eb02

**Comment by frontendschlampe** [comment]
We need a ``composer.phar`` more recent than 24th of september which is still working to pinpoint this.

**Comment by aschempp** [comment]
Mine is from August… can’t help here.

**Comment by dmolineus** [comment]
Started some debugging and compiled some composer.phar at own. It seems to be between:
https://github.com/composer/composer/compare/15face54...0adcab7

If you want to get the composer.phar just tell me. Gonna have a deeper look.

**Comment by dmolineus** [comment]
I tracked it down:
 - The semver classes got extracted to a separate lib.
 - The new constraint class is `Composer\Semver\Constraint/Constraint` but the plugin uses `Composer\Package\LinkConstraint\VersionConstraint`
 - The version constraint extends the new provided class.

Here is the point there it breaks:
https://github.com/composer/semver/blob/master/src/Constraint/Constraint.php#L159

 - There is a comparison of 2 versions accessing the private vars. This works fine as long as both classes are identical but not if one class is a child class of the main one.

Imho it's a bug by composer/semver. It could be fixed in the cca plugin by using the new Constraint class here. But still a fallback for older composer installation should be provided.
https://github.com/contao-community-alliance/composer-plugin/blob/master/src/ContaoCommunityAlliance/Composer/Plugin/Plugin.php#L290

**Comment by fritzmg** [comment]
Is there a workaround? I currently experience the same problem in one installation.

**Comment by dmolineus** [comment]
Disable the autamtic composer.phar update and use an older version of the composer.phar. The last one which works at the moment is commit `15face5432d7b7334db6ac69fac0190971cafa6e`.

**Comment by discordier** [comment]
So how shall we proceed here? Changing to class_exists check and use the new class is enough?

**Comment by dmolineus** [comment]
It think this should be enough. So we don't have to wait until composer fixes the issue itself.

**Comment by discordier** [comment]
Could you please test the hotfix branch on your installations @frontendschlampe, @qzminski?

**Comment by frontendschlampe** [comment]
It's working for me! thx

**Comment by discordier** [comment]
Should be fixed with 2.3.7.
Thanks for testing and providing so much input to all of you.

**Comment by qzminski** [comment]
Works for me as well, good job!

**Comment by Toflar** [comment]
Same here, works for me now :+1:

**Comment by xantippe** [comment]
update at my issue https://github.com/contao-community-alliance/composer-client/issues/266

I had make a downgrade to Contao 3.5.2 per LiveUpdate and after them, I can update the MetaModels Packages... I update per LU to 3.5.4 and I see the same error message!

**Comment by discordier** [comment]
You have to update the plugin only before doing anything else.

**Comment by xantippe** [comment]
Composer-Client wurde vorher aktualisiert und Cache gelöscht - bin bei 
 	contao-community-alliance/composer-client 	0.15.4

**Comment by discordier** [comment]
Das plugin muss aktualisiert werden, nicht der Client. Der Bug war im plugin.

**Comment by xantippe** [comment]
wie aktualisiere ich das Ding?

**Comment by Floxn** [comment]
Ich muss es leider auch auf deutsch machen, sonst wäre es glaube ich unverständlich.

Ich habe im Moment das gleiche Problem. Zu meiner Sachlage:
Ich habe alle Pakete auf einmal aktualisiert. Danach wollte ich eine Extension installieren bei der eine Fehlermeldung auftauchte. Beim prüfen fand ich dann ebenfalls den Core in "composer/vendor" ordner.
Da ich bei diesem System in der Nacht immer ein komplettes Backup machen lasse, habe ich beide einmal miteinander verglichen, im Backup war der Core Ordner noch nicht vorhanden. Er muss also beim aktualisieren der Pakete hinzugekommen sein. 

Version: 3.5.3

Vielleicht hilft euch das weiter.

**Comment by dmolineus** [comment]
Ich habe hier eine Vorgehensweise beschrieben, wie man ein defektes System wieder zum Laufen bringt. https://community.contao.org/de/showthread.php?59625-Composer-installiert-Visitors-nicht&p=387837#post387837

Es kann sein, dass man composer Update hier zweimal anstoßen muss.

**Comment by fritzmg** [comment]
@xantippe @Floxn you could simply download the latest release here on GitHub and then copy the content of the `composer-plugin-2.3.7` folder within the archive to `composer/vendor/contao-community-alliance/composer-plugin` and overwrite the existing content.

However, this will still prevent you from updating the packages, since Composer will detect changes in the source folders. You need to enable the "Discard changes" setting within the Composer settings of the package management.

**Comment by Floxn** [comment]
yes, no it´s working. Thanks for your help!

**Comment by xantippe** [comment]
@fritzmg: sorry, ich bekomm´s nicht hin: https://community.contao.org/de/showthread.php?59459-Composer-versucht-Core-Update-durchzuf%FChren&p=388711#post388711

**Comment by xantippe** [comment]
hmmm :-(
ich habe unter Contao 3.5.4:
* composer-plugin mit Version 2.3.8 von Githug bezogen und ```composer\vendor\contao-community-alliance\composer-plugin``` aktualisiert
* ```composer\cache``` gelöscht
* es kommen die Fehlermeldungen
```
Deprecated notice: The Composer\Package\LinkConstraint\VersionConstraint class is deprecated, use Composer\Semver\Constraint\Constraint instead. in phar://C:/Contao2Go/htdocs/vhosts/contao-3.5.x/composer/composer.phar/src/Composer/Package/LinkConstraint/VersionConstraint.php on line 17

#0 [internal function]: __error(16384, 'The Composer\\Pa...', 'phar://C:/Conta...', 17, Array)
#1 phar://C:/Contao2Go/htdocs/vhosts/contao-3.5.x/composer/composer.phar/src/Composer/Package/LinkConstraint/VersionConstraint.php(17): trigger_error('The Composer\\Pa...', 16384)
#2 C:\Contao2Go\htdocs\vhosts\contao-3.5.x\vendor\composer\ClassLoader.php(412): include('phar://C:/Conta...')
#3 C:\Contao2Go\htdocs\vhosts\contao-3.5.x\vendor\composer\ClassLoader.php(301): Composer\Autoload\includeFile('phar://C:/Conta...')
#4 [internal function]: Composer\Autoload\ClassLoader->loadClass('Composer\\Packag...')
#5 C:\Contao2Go\htdocs\vhosts\contao-3.5.x\system\modules\!composer\src\Controller\InstalledController.php(220): spl_autoload_call('Composer\\Packag...')
#6 C:\Contao2Go\htdocs\vhosts\contao-3.5.x\system\modules\!composer\src\Controller\InstalledController.php(97): ContaoCommunityAlliance\Contao\Composer\Controller\InstalledController->buildGroupedPackagesList(Object(Composer\Package\RootPackage), Object(Composer\Repository\InstalledFilesystemRepository), Array, Array, Array)
#7 C:\Contao2Go\htdocs\vhosts\contao-3.5.x\system\modules\!composer\src\ClientBackend.php(218): ContaoCommunityAlliance\Contao\Composer\Controller\InstalledController->handle(Object(Contao\Input))
#8 C:\Contao2Go\htdocs\vhosts\contao-3.5.x\system\modules\core\classes\Backend.php(423): ContaoCommunityAlliance\Contao\Composer\ClientBackend->generate()
#9 C:\Contao2Go\htdocs\vhosts\contao-3.5.x\system\modules\core\controllers\BackendMain.php(131): Contao\Backend->getBackendModule('composer')
#10 C:\Contao2Go\htdocs\vhosts\contao-3.5.x\contao\main.php(20): Contao\BackendMain->run()
#11 {main}
```
* Composer Plugin aktualisiert - Meldung "Warning: SNI support not available, OpenSSL version too old. Host verification has been deactivated." und
"Composer wurde aktualisiert!"
* MM-Core/Bundle angeklickt + Testlauf 
kommt weiterhin
```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - The requested package contao/core == 3.5.2.0 could not be found.
```
MM kann ich aktualisieren, wenn ich per LU auf 3.5.2 runter gehe - aber das "runter&hoch" der C-Versionen ist nicht "schön"

was nun tun?

**Comment by fritzmg** [comment]
Die Deprecated Notice bekommst du weg, indem du den composer-client auf 0.16.0-beta1 aktualisierst.

Da bei dir der contao/core schon installiert wurde, musst du nun bspw. wie hier vorgehen:
* https://community.contao.org/de/showthread.php?59625-Composer-installiert-Visitors-nicht&p=387837&viewfull=1#post387837
* https://community.contao.org/de/showthread.php?59459-Composer-versucht-Core-Update-durchzuf%FChren&p=388715&viewfull=1#post388715

**Comment by xantippe** [comment]
hatte zwar kein Contao in /composer/vendor aber...

ich habe nun wie in https://community.contao.org/de/showthread.php?59459-Composer-versucht-Core-Update-durchzuf%FChren&p=388725&viewfull=1#post388725 die zwei Ornder gelöscht auf der Konsole composer.phar update ausgeführt

hat eine Menge installiert - aber die Meldung folgte

```
  [ContaoCommunityAlliance\Composer\Plugin\DuplicateContaoException]
  Warning: Contao core was about to get installed but has been found in proje
  ct root, to recover from this problem please restart the operation
```

**Comment by MacKP** [comment]
@xantippe wie weiter oben beschrieben dann einfach noch mal machen. Dann sollte das durchlaufen. Hatte ich auch.

**Comment by xantippe** [comment]
o.k. - hat funktioniert...

**Comment by JLBA** [comment]
Hi everyone,

Don't know if you fixed the problem but I'm experiencing it in a Contao 3.5.8 local installation (MAMP). 

And another strange message : 
```
$ /usr/bin/env php -d memory_limit=1G -d max_execution_time=3000 composer.phar update --no-ansi --no-interaction -v
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - league/uri 4.1.x-dev requires php >=5.5.9 -> your PHP version (5.4.45) does not satisfy that requirement.
    - league/uri 4.1.0 requires php >=5.5.9 -> your PHP version (5.4.45) does not satisfy that requirement.
    - league/uri 4.0.1 requires php >=5.5.9 -> your PHP version (5.4.45) does not satisfy that requirement.
    - league/uri 4.0.0 requires php >=5.5.9 -> your PHP version (5.4.45) does not satisfy that requirement.
    - contao/core-bundle 4.2.x-dev requires php >=5.5.0 -> your PHP version (5.4.45) does not satisfy that requirement.
    - lionel/coin-slider 2.0.0-beta2 requires contao/core-bundle ~4.0 -> satisfiable by contao/core-bundle[4.0.0, 4.0.0-RC1, 4.0.0-beta1, 4.0.1, 4.0.2, 4.0.3, 4.0.4, 4.1.0, 4.1.0-RC1, 4.1.0-beta1, 4.1.1, 4.2.x-dev, 4.0.x-dev].
    - lionel/coin-slider 2.0.0-beta requires contao/core-bundle ~4.0 -> satisfiable by contao/core-bundle[4.0.0, 4.0.0-RC1, 4.0.0-beta1, 4.0.1, 4.0.2, 4.0.3, 4.0.4, 4.1.0, 4.1.0-RC1, 4.1.0-beta1, 4.1.1, 4.2.x-dev, 4.0.x-dev].
    - lionel/coin-slider 2.0.0 requires contao/core-bundle ~4.0 -> satisfiable by contao/core-bundle[4.0.0, 4.0.0-RC1, 4.0.0-beta1, 4.0.1, 4.0.2, 4.0.3, 4.0.4, 4.1.0, 4.1.0-RC1, 4.1.0-beta1, 4.1.1, 4.2.x-dev, 4.0.x-dev].
    - Can only install one of: contao/core-bundle[4.0.x-dev, 3.5.8].
    - contao/core 3.5.8 requires contao/core-bundle 3.5.8.0 -> satisfiable by contao/core-bundle[3.5.8].
    - contao/core-bundle 4.0.0 conflicts with contao/core[3.5.8].
    - contao/core-bundle 4.0.0-RC1 conflicts with contao/core[3.5.8].
    - contao/core-bundle 4.0.0-beta1 conflicts with contao/core[3.5.8].
    - contao/core-bundle 4.0.1 conflicts with contao/core[3.5.8].
    - contao/core-bundle 4.0.2 conflicts with contao/core[3.5.8].
    - contao/core-bundle 4.0.3 conflicts with contao/core[3.5.8].
    - contao/core-bundle 4.0.4 conflicts with contao/core[3.5.8].
    - contao/core-bundle 4.1.0 requires league/uri ~4.0 -> satisfiable by league/uri[4.0.0, 4.0.1, 4.1.0, 4.1.x-dev].
    - contao/core-bundle 4.1.0-RC1 requires league/uri ~4.0 -> satisfiable by league/uri[4.0.0, 4.0.1, 4.1.0, 4.1.x-dev].
    - contao/core-bundle 4.1.0-beta1 requires league/uri ~4.0 -> satisfiable by league/uri[4.0.0, 4.0.1, 4.1.0, 4.1.x-dev].
    - contao/core-bundle 4.1.1 requires league/uri ~4.0 -> satisfiable by league/uri[4.0.0, 4.0.1, 4.1.0, 4.1.x-dev].
    - Installation request for contao/core 3.5.8 -> satisfiable by contao/core[3.5.8].
    - Installation request for lionel/coin-slider >=2.0.0.0,<2.1-dev -> satisfiable by lionel/coin-slider[2.0.0, 2.0.0-beta, 2.0.0-beta2].
```
Why does-it ask for contao/core-bundle 4.x.x ?

**Comment by aschempp** [comment]
@JLBA no you're not. The error message you posted earlier tells me you're trying to install `lionel/coin-slider 2.0.0` which is only compatible with Contao 4.

This issue has been fixed in version 2.5

**Comment by aschempp** [comment]
Sorry, version 2.4 I mean

**Comment by JLBA** [comment]
Ok. Thanks.
JL Test
jltest@free.fr
